sqlx-mysql/src/connection/infile.rs

```diff
@@ -0,0 +1,298 @@
+    pub fn get_writer<'a>(&'a mut self) -> InfileWriter<'a> {
+        InfileWriter::new(&mut self.conn.stream)
+    }
+
+    /// Get the filename that MySql requested from the LOCAL INFILE
+    pub fn get_filename(&self) -> &[u8] {
+        &self.filename
+    }
+
+    /// Send data to the database.
+    ///
+    /// The data is buffered and send to the server in packets of at most 16MB. The data is automatically flushed when the buffer is full.
+    ///
+    /// ### Note: Completion Step Required
+    /// You must still call [finish()](Self::finish) to complete the process.
+    pub async fn send(&mut self, source: impl Deref<Target = [u8]>) -> Result<(), Error> {
+        let mut right = source.deref();
+        while !right.is_empty() {
+            let (left, right_) = right.split_at(std::cmp::min(MAX_MYSQL_PACKET_SIZE, right.len()));
+            self.buf.extend_from_slice(left);
+            if self.buf.len() >= MAX_MYSQL_PACKET_SIZE + 4 {
+                assert_eq!(self.buf.len(), MAX_MYSQL_PACKET_SIZE + 4);
+                self.drain_packet(MAX_MYSQL_PACKET_SIZE).await?;
+                assert!(self.buf.is_empty());
+                self.buf.extend_from_slice(&[0; 4]);
+            }
+            right = right_;
+        }
+        Ok(())
+    }
+
+    /// Flush the stream.
+    pub async fn flush(&mut self) -> Result<(), Error> {
+        if self.buf.len() > 4 {
+            // Cannot have multiple packets in buffer, as they would have been drained by write() already
+            assert!(self.buf.len() <= MAX_MYSQL_PACKET_SIZE + 4);
+            self.drain_packet(self.buf.len() - 4).await?;
+        }
+        Ok(())
+    }
+
+    async fn drain_packet(&mut self, len: usize) -> Result<(), Error> {
+        self.buf[0..3].copy_from_slice(&(len as u32).to_le_bytes()[..3]);
+        self.buf[3] = self.conn.stream.sequence_id;
+        self.conn
+            .stream
+            .socket
+            .socket_mut()
+            .write(&self.buf[..len + 4])
+            .await?;
+        self.buf.drain(..len + 4);
+        self.conn.stream.sequence_id = self.conn.stream.sequence_id.wrapping_add(1);
+        Ok(())
+    }
+
+    /// Finish sending the LOCAL INFILE data to the server.
+    ///
+    /// This must always be called after you're done writing the data.
+    /// Returns the number of rows that were inserted.
+    pub async fn finish(mut self) -> Result<u64, Error> {
+        self.flush().await?;
+        self.conn.stream.send_empty_response().await?;
+        let packet = self.conn.stream.recv_packet().await?;
+        let packet = packet.ok()?;
+        Ok(packet.affected_rows)
+    }
+}
+
+/// A writer that writes to a [`MySqlLocalInfile`] stream.
+pub struct InfileWriter<'a> {
+    stream: &'a mut MySqlStream,
+    send: Option<SendPacket>,
+}
+
+impl<'a> InfileWriter<'a> {
+    fn new(stream: &'a mut MySqlStream) -> Self {
+        Self { stream, send: None }
+    }
+}
+
+impl<'a> AsyncWrite for InfileWriter<'a> {
```

Comment:
It's not strictly a blocker, but I'd like to see an implementation of [`tokio::io::AsyncWrite`](https://docs.rs/tokio/latest/tokio/io/trait.AsyncWrite.html) for Tokio users as well.

Reply:
Makes sense, I think that's very doable

Reply:
This is now implemented in [8d4a839](https://github.com/launchbadge/sqlx/pull/2738/commits/8d4a839bf7da769c6951603788da39eed326cb0b)